src/icu_win.cc

```diff
@@ -0,0 +1,238 @@
+    }
+  }
+
+  // Our goal here is to load the ICU DLL from the system directory, even when
+  // the current process has loaded "icu.dll" from somewhere other than the
+  // system directory. To do so we must use the full path here.
+  const std::wstring icu_dll_path = GetSystem32Dir() + L"\\icu.dll";
+
+  // CAVEAT: LoadLibraryExW with LOAD_LIBRARY_SEARCH_SYSTEM32 is not
+  // sufficient when "icu.dll" is already loaded from somewhere other than
+  // the system directory. This is why we must pass a full path here.
+  const HMODULE icu_dll = ::LoadLibraryW(icu_dll_path.c_str());
+  if (icu_dll == nullptr) {
+    g_unavailable.store(true, std::memory_order_relaxed);
+    return Unavailable();
+  }
+
+  const auto ucal_getHostTimeZoneRef =
+      AsProcAddress<ucal_getHostTimeZone_func>(icu_dll, "ucal_getHostTimeZone");
+  const auto ucal_getTimeZoneIDForWindowsIDRef =
+      AsProcAddress<ucal_getTimeZoneIDForWindowsID_func>(
+          icu_dll, "ucal_getTimeZoneIDForWindowsID");
+
+  // Note: ucal_getHostTimeZone can be unavailable on older Windows.
+  if (!ucal_getTimeZoneIDForWindowsIDRef) {
+    g_unavailable.store(true, std::memory_order_relaxed);
+    return Unavailable();
+  }
+
+  // Store the function pointers
+  g_ucal_getHostTimeZone.store(ucal_getHostTimeZoneRef,
+                               std::memory_order_relaxed);
+  g_ucal_getTimeZoneIDForWindowsID.store(ucal_getTimeZoneIDForWindowsIDRef,
+                                         std::memory_order_relaxed);
+
+  return {true, ucal_getHostTimeZoneRef, ucal_getTimeZoneIDForWindowsIDRef};
+}
+
+// Convert wchar_t array (UTF-16) to UTF-8 string
+std::string Utf16ToUtf8(const wchar_t* ptr, size_t size) {
+  if (size > std::numeric_limits<int>::max()) {
+    return std::string();
+  }
+  const int chars_len = static_cast<int>(size);
+  int num_bytes_in_utf8 = 0;
+  {
+    std::array<char, 32> buffer;
+    num_bytes_in_utf8 =
+        ::WideCharToMultiByte(CP_UTF8, WC_ERR_INVALID_CHARS, ptr, chars_len,
+                              buffer.data(), buffer.size(), nullptr, nullptr);
+    if (num_bytes_in_utf8 <= buffer.size()) {
+      return std::string(buffer.data(), num_bytes_in_utf8);
+    }
+  }
+  auto buffer = std::unique_ptr<char[]>(new char[num_bytes_in_utf8]);
+  const int num_written_bytes =
+      ::WideCharToMultiByte(CP_UTF8, WC_ERR_INVALID_CHARS, ptr, chars_len,
+                            buffer.get(), num_bytes_in_utf8, nullptr, nullptr);
+  if (num_written_bytes != num_bytes_in_utf8) {
+    return std::string();
+  }
+  return std::string(buffer.get(), num_written_bytes);
+}
+
+}  // namespace
+
+std::string GetWindowsLocalTimeZone() {
+  const auto icu = GetIcuFunctions();
+  if (!icu.available) {
+    return std::string();
+  }
+
+  std::int32_t length = 0;
+
+  UErrorCode status = U_ZERO_ERROR;
+
+  // Try ucal_getHostTimeZone first (available on Windows 11+)
+  if (icu.ucal_getHostTimeZone != nullptr) {
+    std::array<UChar, 32> buffer;
```

Comment:
Is "32" some well-calculated value for the length of a timezone name (in which case it probably deserves a name and a comment), or a guess?

Reply:
This is a guess. Added a comment for future readers.